spec.md

```diff
@@ -0,0 +1,81 @@
+
+**Trace Attributes** are key/value pairs stored in a Trace Context and propagated _in-band_ to all future child Spans. Given a full-stack OpenTracing integration, Trace Attributes enable powerful functionality by transparently propagating arbitrary application data: e.g., from a mobile app all the way into the depths of a storage system. Trace attributes come with powerful _costs_ as well; since the attributes are propagated in-band, if they are too large they can reduce system throughput or contribute to RPC latency. Use Trace Attributes with care.
+
+**Trace Attributes** vs. **Span Tags**
+
+* Trace Attributes are propagated in-band across process boundaries. Span Tags are not propagated.
+* Span Tags are recorded off-band in the tracing system's storage. Trace Attributes are not recorded.
+
+## Platform-Independent API Semantics
+
+OpenTracing supports a number of different platforms, and of course the per-platform APIs try to blend in to their surroundings and do as the Romans do. That said, each platform API must model a common set of semantics for the core tracing concepts described above. In this document we attempt to describe those concepts and semantics in a language- and platform-agnostic fashion.
+
+#### Tracer
+
+The `Tracer` interface must have the following capabilities:
+
+- Start a `Span` that has no parent **(py: `start_trace`, go: `StartTrace`)**
+- Start a `Span` given a parent `TraceContext` **(py: `join_trace`, go: `JoinTrace`)**
+- Start a `Span` associated with a specific `TraceContext` **(py: `start_span_with_context`, go: `StartSpanWithContext`)**
+
+
+#### Span
+
+The `Span` interface must have the following capabilities:
+
+- Create and start a new child `Span` with a given operation name **(py: `start_child`, go: `StartChild`)**
+- Finish the (already-started) `Span`.  Finish should be the last call made to any span instance, and to do otherwise leads to undefined behavior. **(py: `finish`, go: `Finish`)**
+- Set a key:value tag on the `Span`. The key must be a `string`, and the value must be either a `string`, a `boolean`, or a numeric type. Behavior for other value types is undefined. If multiple values are set to the same key (i.e., in multiple calls), implementation behavior is also undefined. **(py: `set_tag`, go: `SetTag`)**
+- Add a new info `Log` to the `Span`, accepting a message `string` and 0 or more payload arguments. The payload arguments may be of any type and arbitrary size, though implementations are not required to retain all payload arguments (or even all parts of all payload arguments). **(py: `info`, go: `Info`)**
+- Add a new error `Log` to the `Span`; analagous to adding an info log aside from the severity. **(py: `error`, go: `Error`)**
+- Access the `TraceContext` associated with the `Span`. **(py: `trace_context`, go: `TraceContext()`)**
+
+
+#### TraceContext
```

Comment:
@dankosaur You're right, Dan, attributes are the equivalent of "baggage" (let me use this term for disambiguation), and they are completely optional and not required for trace/span ID. But they are essential for "baggage" propagation if the developers have a need for that. At my company a lot of people are actually very excited about "baggage", planning to use that for security credentials propagation, and for resource usage cost accounting.

That's why my preference is to group/separate functionality discussion into
- propagation functionality
    - trace/span ID
    - baggage
- serialization functionality

NB: we had a side discussion elsewhere that propagation as a topic might be more complex, not just "to all descendants", but "direct children only", or even "ancestors".